File: backend/src/routes/files.js
import { Router } from "express";
import { requireAuth } from "../middleware/auth.js";
import { initiateUpload, processFile, listFiles, deleteFile } from "../services/fileService.js";

const router = Router();

router.post("/upload", requireAuth, async (req, res) => {
  try {
    const { filename, content_type } = req.body;
    if (!filename || !content_type) {
      return res.status(400).json({ detail: "filename and content_type are required" });
    }
    const result = await initiateUpload(req.userId, filename, content_type);
    return res.status(201).json(result);
  } catch (err) {
    console.error("Upload error:", err);
    return res.status(err.status || 500).json({ detail: err.message });
  }
});

router.post("/:fileId/process", requireAuth, async (req, res) => {
  try {
    const result = await processFile(req.params.fileId, req.userId);
    return res.json(result);
  } catch (err) {
    console.error("Process error:", err);
    return res.status(err.status || 500).json({ detail: err.message });
  }
});

router.delete("/:fileId", requireAuth, async (req, res) => {
  try {
    const result = await deleteFile(req.params.fileId, req.userId);
    return res.json(result);
  } catch (err) {
    console.error("Delete file error:", err);
    return res.status(err.status || 500).json({ detail: err.message });
  }
});

router.get("/", requireAuth, async (req, res) => {
  try {
    const files = await listFiles(req.userId);
    return res.json(files);
  } catch (err) {
    console.error("List files error:", err);
    return res.status(err.status || 500).json({ detail: err.message });
  }
});

export default router;

File: backend/app/api/routes/files.py
from fastapi import APIRouter, Depends, HTTPException, status

from app.domain.entities import FileResponse, UploadRequest, UploadResponse
from app.domain.exceptions import FileProcessingError, NotFoundError
from app.services.file_service import FileService
from app.api.deps import get_current_user, get_file_service

router = APIRouter(prefix="/files", tags=["files"])


@router.post("/upload", response_model=UploadResponse, status_code=status.HTTP_201_CREATED)
def upload_file(
    body: UploadRequest,
    user_id: str = Depends(get_current_user),
    file_service: FileService = Depends(get_file_service),
):
    file_id, upload_url = file_service.initiate_upload(
        user_id=user_id,
        filename=body.filename,
        content_type=body.content_type,
    )
    return UploadResponse(file_id=file_id, upload_url=upload_url)


@router.post("/{file_id}/process", response_model=FileResponse)
async def process_file(
    file_id: str,
    user_id: str = Depends(get_current_user),
    file_service: FileService = Depends(get_file_service),
):
    try:
        result = await file_service.process_file(file_id=file_id, user_id=user_id)
        return FileResponse(
            file_id=result.file_id,
            original_filename=result.original_filename,
            file_type=result.file_type,
            status=result.status,
            upload_date=result.upload_date,
            transaction_count=result.transaction_count,
        )
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=e.message)
    except FileProcessingError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=e.message)


@router.get("", response_model=list[FileResponse])
def list_files(
    user_id: str = Depends(get_current_user),
    file_service: FileService = Depends(get_file_service),
):
    files = file_service.get_user_files(user_id)
    return [
        FileResponse(
            file_id=f.file_id,
            original_filename=f.original_filename,
            file_type=f.file_type,
            status=f.status,
            upload_date=f.upload_date,
            transaction_count=f.transaction_count,
        )
        for f in files
    ]
